cargo-dist/src/lib.rs

```diff
@@ -666,6 +692,129 @@ fn gather_work(cfg: &Config) -> Result<DistGraph> {
         }
     }
 
+    // Add release notes
+    for ((_app_name, version), (package_id, release)) in &mut releases {
+        let package_info = &workspace.package_info[&&*package_id];
+
+        // Try to parse out relevant parts of the changelog
+        // FIXME: ...this is kind of excessive to do eagerly and for each crate in the workspace
+        if let Some(changelog_path) = &package_info.changelog_file {
+            if let Ok(changelog_str) = try_load_changelog(changelog_path) {
+                let changelogs = parse_changelog::parse(&changelog_str)
+                    .into_diagnostic()
+                    .wrap_err_with(|| format!("failed to parse changelog at {changelog_path}"));
+                if let Ok(changelogs) = changelogs {
+                    let version_string = format!("{}", package_info.version);
+                    if let Some(release_notes) = changelogs.get(&*version_string) {
+                        release.changelog_title = Some(release_notes.title.to_owned());
+                        release.changelog_body = Some(release_notes.notes.to_owned());
+                    }
+                }
+            }
+        }
+
+        use std::fmt::Write;
+        let mut changelog_body = String::new();
+
+        let mut installers = vec![];
+        let mut bundles = vec![];
+        let mut symbols = vec![];
+        let mut other = vec![];
+        for artifact_idx in &release.artifacts {
+            let artifact = &artifacts[artifact_idx.0];
+            match artifact.kind {
+                ArtifactKind::ExecutableZip => {
+                    bundles.push(artifact);
+                }
+                ArtifactKind::Symbols => {
+                    symbols.push(artifact);
+                }
+                ArtifactKind::DistMetadata => {
+                    // Do nothing
+                }
+                ArtifactKind::Installer => {
+                    installers.push(artifact);
+                }
+                ArtifactKind::Unknown => {
+                    other.push(artifact);
+                }
+                _ => todo!(),
+            }
+        }
+
+        if !installers.is_empty() {
+            changelog_body.push_str("## Install\n\n");
+            for installer in installers {
+                let install_hint;
+                let description;
+
+                match &installer.bundle {
+                    BundleStyle::Installer(InstallerImpl::GithubShell(info)) => {
```

Comment:
this function is Very Large, it may be valuable to break it up into smaller chunks